### ISSUE TITLE
Autofocus performance degraded when there are a lot of saturated pixels

### ISSUE THREAD

**Issue by ericweiner** [EtalumaSupport/LumaViewPro#319]
The AF algorithm does not perform very well when there are a lot of saturated pixels.  Since those are not providing any information, perhaps we can ignore those pixels in the algorithm?  Or find another method to improve performance?

**Comment by ericweiner** [comment]
## Triage / analysis (no fix yet)

**Disposition:** fix-incomplete (unconfirmed)

**Root cause:** a symptom-masking fix (`c55d07b9`) shipped but is unverified (no post-fix saturated-scene log). The Vollath F4 metric (`autofocus_functions.py:60-71`) has no saturation handling, but "AF scores must be stable and unsaturated" has no single owner -- it is split across a metric-level mask (#319), a missing AG-disable in the runner (#701), and a watchdog (#703). The shipped mask has hazards (zero-injection edges; dtype-max threshold never fires on uint16/Mono12; no test).

**Proposed fix shape:** `AutofocusRunner`/`ImagingAPI` establishes a frozen, unsaturated acquisition context before scoring; the focus function stays a pure metric -- awaiting plan approval.

**Cluster:** shares root with #701 (likely the true primary -- AG generates the saturation) and #703.

Status: analyzed; needs a post-fix saturated-scene AF log with score logging; fix not started.

**Comment by ericweiner** [comment]
Possible mechanism identified for the 12-bit case: #730 -- the autofocus saturated-pixel mask (`_mask_saturated`, modules/autofocus_functions.py) thresholds on the container dtype max (65535 for uint16), so on 12-bit frames (max 4095) saturated pixels are never masked and DO dominate focus scores, which is this issue's reported symptom. Fix shape + sequencing recorded on #730.

**Comment by ericweiner** [comment]
## Branch-tip verification (fix/merge-review-clusters @ 213ba1bb): NOT fixed on this branch

Checked at the merge-review batch tip. `modules/autofocus_functions.py` is byte-identical
to shipping (4.0.0-beta @ e9b76186) -- untouched by the batch. `_mask_saturated`
(autofocus_functions.py:38) still thresholds on `np.iinfo(image.dtype).max` (65535 on the
uint16 container), so on 12-bit frames (full scale 4095) the mask never fires and saturated
pixels still dominate focus scores. This is the mechanism #730 named, and it remains OPEN.

Note the batch DID fix a sibling site sharing this root: the capture-evidence /
blown-frame saturation metric (`_saturated_fraction` in lumascope_api/imaging.py) now
derives full scale from the frame-stamped `significant_bits` (Cluster 6, 9377ac51). That is
#730 site 2, not the AF mask (#730 site 1). The AF-mask member was deliberately deferred
(changes AF scores on 12-bit -> needs a bench re-check).

Two-part status:
- 12-bit masking mechanism (#730 site 1): NOT fixed on branch. One-line fix designed
  (threshold on `(1 << significant_bits) - 1`, plus plumbing the frame depth into
  focus_function/_mask_saturated) -- pending post-shipping-merge scheduling + a 12-bit AF
  bench pass.
- The original ask (AF degraded on saturated scenes generally, incl. 8-bit): still an open
  algorithm-quality request. The existing shipped mask (c55d07b9) is a symptom-mask, still
  unverified (no post-fix saturated-scene AF score log) and untested. Ownership stays split
  with #701 (AG-disable, likely the true primary) and #703 (watchdog).

No regression test exercises `_mask_saturated`. Chronology gate clean (no reporter comment
postdates any fix). Keep OPEN; keep the Analyzed label.

**Comment by ericweiner** [comment]
## Correction: the "12-bit masking blindness" mechanism is falsified -- AF scores 8-bit frames

The earlier analysis named 12-bit mask blindness (`_mask_saturated` thresholding on the uint16 container max) as a cause of AF degrading on saturated scenes. Tracing the AF capture path falsifies that mechanism:

- The AF sweep captures via `capture_and_wait` with the default `force_to_8bit=True`, so `_mask_saturated` scores a uint8 frame. `convert_to_8bit` maps a saturated 12-bit pixel to 255, so the dtype-max threshold catches saturation correctly. See the companion correction on #730 for the full trace.

So there is no 12-bit-threshold fix to make for this issue. If AF still degrades on heavily-saturated scenes, the cause lives elsewhere -- most likely auto-gain driving the saturation in the first place (#701), plus the edge hazard of the existing zero-injection mask altering the Vollath autocorrelation at masked-region boundaries. Those are behavior-changing and bench-coupled; they are the real owners of "AF scores stable + unsaturated," not a depth threshold.

Keeping Analyzed. The original "improve AF on saturated scenes" ask remains open and is split with #701 (auto-gain disable during AF) and #703 (watchdog).

**Comment by ericweiner** [comment]
**2026-07-08 currency sweep (LVP 4.0.0-beta @ `5b714259`) -- STILL PRESENT**

Partial only. A saturated-pixel mask shipped (`c55d07b9`, `_mask_saturated` at `modules/autofocus_functions.py:35`) and fires on the 8-bit AF frames, but it is untested and does not resolve the issue: the likely primary is auto-gain saturating the scene (owned by #701/#703), and the zero-mask can perturb the Vollath F4 score. Keep open. Next: a bench post-fix saturated-scene AF-score log + a `_mask_saturated` regression test + ownership resolution with #701/#703.

**Comment by ericweiner** [comment]
## Prior art (2026-07-23 doc cross-reference)

**Governing doc:** None -- No governing plan/audit -- genuine coverage gap. Root: auto-gain drifting brightness during the sweep.
**What this means:** NONE -- Clusters with #701/#703/#319 -- wants one plan, not three tickets.
**Cluster:** #701, #703 (AF robustness under auto-gain).

Read the governing doc before re-analyzing this issue.

**Comment by ericweiner** [comment]
## Cluster refinement (adversarial pass 2026-07-28)

Membership in the AF acquisition-context cluster (#701's freeze/restore owner) is CONDITIONAL: the auto-gain link is hedged on this issue's own record, an intrinsically bright sample saturates under frozen gain too, and the `c55d07b9` mask-boundary hazard (`modules/autofocus_functions.py:35`) survives a freeze regardless. The 12-bit-mask hypothesis remains falsified (AF frames forced 8-bit). One plan across #701/#319/#703 stands; this issue keeps its own metric/mask limb inside it.

Status: analyzed (unchanged); fix not started.

**Comment by ericweiner** [comment]
## Currency re-check at `f78bf007`: NOTHING CHANGED. Plus the #753 ruling, which postdates every annotation here.

*(Symbols named; line numbers bound to `f78bf007`.)*

| Code fact | Verdict |
|---|---|
| `_mask_saturated` thresholds on the container dtype max | **STILL TRUE** -- `np.iinfo(image.dtype).max`; `modules/autofocus_functions.py` has **zero** commits since `1b83a3e8` (2026-07-06) |
| AF sweep still grabs 8-bit (so the 12-bit falsification holds) | **STILL TRUE in mechanism, CHANGED in spelling** -- see below |
| Vollath F4 has no saturation handling of its own | **STILL TRUE** |
| No regression test exercises `_mask_saturated` | **STILL TRUE** -- `grep -rln mask_saturated tests/` returns nothing |

**One citation correction for future readers:** `d2f751ac` (2026-08-19) renamed the grab.
`AutofocusRunner` now calls `imaging._capture_and_wait_impl(...)`, not `capture_and_wait`.
**Every prior annotation on this issue cites the stale name.** The mechanism is unchanged --
the parameter is still `force_to_8bit: bool = True` and the runner passes no override
(`grep -c force_to_8bit modules/autofocus_runner.py` -> `0`), so `_mask_saturated` still
scores a uint8 frame and **the 12-bit hypothesis stays falsified.** Do not resurrect it.

## The #753 ruling: PARTIAL coverage -- it takes the auto-gain limb, not the metric limb

Ruling (c) (bounded converge-then-lock at scan start, per layer, loud typed failure,
frozen values into metadata) lands on the **right seam** for this issue's auto-gain half.
A premise worth correcting explicitly, because it was wrong when this re-check was briefed:

**Standalone AF and "AF All Steps" are NOT off the protocol path.** All three AF entry
points -- `VerticalControl.run_autofocus_from_ui`,
`ProtocolSettings.run_autofocus_scan_from_ui`, and the protocol step runner -- drive the
same `SequencedCaptureRunner` (the first two pass
`run_mode=SequencedCaptureRunMode.SINGLE_AUTOFOCUS_SCAN`). So a scan-entry lock does reach
standalone AF.

Also verified: **nothing disarms AG at run start.** `AutofocusRunner` contains zero
`auto_gain` references; the runner's bring-up only calls
`_update_auto_gain_target_brightness_impl` -- a *target* write, not a disarm -- while live
view arms hardware AG through `_apply_layer_camera_settings_impl`. **Continuous AG during
AF is inherited from live view at scan entry.**

**Consequence for the #753 plan (Rule 35):** (c) makes #701's separate "AF-side auto-gain
disable" redundant -- #701 should not build a second mechanism. The one decision worth
stating explicitly in that plan, cheap now and expensive to retrofit: **does the lock arm
for `run_mode=SINGLE_AUTOFOCUS_SCAN`, or only for the saving protocol modes?**

## What ruling (c) does NOT cover -- this issue's own limb survives

1. Frozen-but-saturating gain still saturates. Locking the value does not unsaturate a
   bright sample.
2. The `_mask_saturated` zero-injection edge hazard perturbs the Vollath autocorrelation at
   masked-region boundaries, and **survives any freeze**.
3. The mask is still **untested**.
4. Ruling condition 2 (frozen values into image metadata) **does not reach AF** -- both AF
   scan plans set `enable_image_saving=False`, so there is no image to stamp.

## Chronology

`git log --all --grep='#319\b'` returns nothing -- **no fix SHA has ever existed**. Gate is
vacuously clean.

**Still unmeasured, and it is the gap that matters:** the symptom has never been re-measured.
No post-`c55d07b9` saturated-scene AF score log exists, and the edge-hazard magnitude is a
mechanism argument, not a measurement.

Status: **analyzed, UNCHANGED.** Keep OPEN, keep `Analyzed`. Fix not started.

**Comment by ericweiner** [comment]
## Triage record: a remedy exists on paper, was never ratified, and this issue's scope is contested

From a re-triage pass run against `4.0.0-beta` at `f78bf007`.

**The remedy.** `LVP_STATE_OWNERSHIP.md:82` prescribes the owner for the state this issue turns on -- the same row family that governs #701 -- and `AUDIT_SHIPPING_MERGE_REVIEW_2026-07-02.md:139` carries the adoption claim.

**Its status, stated precisely, because it matters.** That document describes its own rows as adopted, but the phrase "Adopt this owner map" appears **only inside its own disposition cell** -- an unfiltered search of the corpus returns that line and one report quoting it, nothing else. The contemporaneous record credits Eric with choosing the *method* and says "Eric to confirm" and "Phase-3 fix planning is gated on Eric's approval." Git authorship settles nothing either way, since sessions commit under Eric's identity.

So: **a remedy was proposed and never ratified by a named person.** Treat it as input to a plan, not as an approved design. Nothing has been built -- `resolve_step_z`, the resolver the same row family depends on, returns zero hits on every branch.

**Scope tension, reported and not resolved.** `EXCLUDED.tsv:8` lists this issue as `PRE-2026`, out of the re-triage scope. `BACKLOG_CLUSTERS_AND_ORDER_2026-09-02.md:85` nevertheless lists it as a cluster member, and that document records the collision against itself at `:21-22`:

> "**#319 is in root 4's membership but is EXCLUDED from the re-triage as PRE-2026**, so root 4 can never be declared complete without a decision on it."

Precedent exists for overriding the exclusion: #524, also marked PRE-2026, was triaged anyway without a recorded decision.

Status: analyzed; remedy proposed but unratified; scope membership undecided; nothing built.

**Comment by ericweiner** [comment]
## Correction to my previous comment: one citation was wrong, and removing it strengthens the point

My earlier comment said `AUDIT_SHIPPING_MERGE_REVIEW_2026-07-02.md:139` "carries the adoption claim". **That is false and I withdraw it.** I relayed it without opening the file. Checked now:

- The file is at `docs/completed/`, not `docs/`.
- Line `:139` is about **#730** and the IDS driver's `significant_bits` pin -- unrelated to this issue.
- The file contains **zero** occurrences of "adopt" and **zero** occurrences of `#319`.

**What survives, re-verified.** `LVP_STATE_OWNERSHIP.md:82` *is* this issue's row -- confirmed by matching each row to its issue rather than by position: `:81` "AG/AE camera MODE during AF" is #701, `:82` "AF metric + saturation/dark conditioning" is this issue, `:83` "AF stuck/progress + best-Z" is #703.

**And the correction makes the original point stronger, not weaker.** The claim was that the owner-map row is a proposal that was never ratified by anyone, with `"Adopt this owner map"` appearing only inside that document's own disposition cell (`:18`). The audit file was the one piece of *external* corroboration I had named for the adoption. It does not corroborate it -- it does not mention it at all. So the row's status is: **proposed, and unratified anywhere in the corpus.**

Unchanged from before: nothing has been built. `resolve_step_z`, the resolver that row family depends on, returns zero hits on every branch.

I am flagging my own error rather than quietly editing, since the thread is the record.

**Comment by ericweiner** [comment]
**Cluster link -- group L3, LINKED.** Other members: #701 #703. Posted to every member so the grouping does not live only in a build directory.

## LINKED group L3 (cross-issue pattern pass, 2026-09-03, pin `47a98931` = beta32): Autofocus owns neither the acquisition context nor a bound for its sweep

Produced by a cross-issue pattern pass over the 124 open issues carrying an analysis, then an adversarial pass over that pass. **No fix started, no plan approved, no status change on any member.** Triage ends at the root cause.

**Members: #319, #701, #703** -- 3 issues, all ANALYZED, #701/#703 `customer facing`.

**Root.** "An AF sweep must run under conditions AF controls" has no owner. The runner does not
freeze the camera's auto-modes (so brightness moves under the metric), the metric has no honest
full-scale (so saturated pixels are not excluded on native-depth frames), and the only wall-clock
bound lives in a Kivy `Clock` closure in the GUI starter.

**Owner -- EXISTS.** `modules/autofocus_runner.py::AutofocusRunner` for the sweep contract;
`modules/lumascope_api/imaging.py::ImagingAPI` for the save/restore it calls. No CREATE.

**Why one change is not enough.**
- **#701** -- `save_camera_state` snapshots gain/exposure **values**, not the auto **mode**; the
  runner never disables AG. Fix: freeze auto-modes for the sweep.
- **#703** -- re-derived at the pin, unfiltered: `AF_SAFETY_TIMEOUT_S` appears in exactly **two
  places, both in `ui/vertical_control.py`** (the constant and its single `Clock.schedule_once`
  arming). The protocol engine's per-step AF takes a different path -- a bare poll-and-return on
  `p._af_future` in `modules/protocol_step_runner.py` -- and is bounded by **nothing**; the only
  bounded wait on that future anywhere is in cleanup, after the run is already ending. Two consumers
  failing in **opposite directions** from one missing owner: too aggressive on one path, unbounded on
  the other. Fix: an API-owned deadline.
- **#319** -- `modules/autofocus_functions.py::_mask_saturated` thresholds on `np.iinfo(dtype).max`.
  Its comment now states the precondition explicitly (both callers pass 8-bit frames), so the site is
  *documented*, not fixed; a native-depth caller would slip past. Fix: the metric receives a real
  full scale, or the runner guarantees the precondition.
Three different changes, three different layers.

**Stakes.** #703 reverts a nearly-focused stage to out-of-focus in front of the operator. #701 makes
every AF score untrustworthy on an auto-gain layer. Note the measured correction on #509: #703's
fixed cost is **once per RUN, not once per AF** -- roughly 8 s of ~864 AF-seconds on a 96-well 3x3
protocol (~0.2%). Anyone ordering these for speed would be ordering the wrong one.

---

## Adversarial pass verdict on the above

**CONFIRMED.** `save_camera_state` snapshots only `gain_db` and `exposure_ms`, never the auto *mode*, while `set_auto_gain` / `set_auto_exposure_time` exist unsnapshotted. `AF_SAFETY_TIMEOUT_S` = exactly 2 occurrences, both in `ui/vertical_control.py`, unfiltered across every extension. The per-step AF waits in `ProtocolStepRunner.scan_iterate` are bare poll-and-return with no timeout, inside a `scan_loop` with no elapsed-time bound; the only bounded wait is `_af_future.exception(timeout=5.0)` in `run_cleanup` — and it waits on `autofocus_thread.current_future`, not on `p._af_future`, which makes the "opposite directions" finding *stronger* than stated. `_mask_saturated` has **one** direct caller; the comment's "both callers" means the two callers of `focus_function`, and both do pass 8-bit.

---

**Provenance of this comment.** The group, its root and its owner were derived by the pattern pass at the pin and then attacked by an independent adversarial pass; both are recorded above, including where the second refuted the first. Foreground re-derivation in this session covered the pixel-size, auto-gain and run-completion groups directly; **this group's claims rest on the two passes rather than on a third foreground check**, so treat any single count here as re-derivable rather than settled. Every prior annotation, including this one, is a hypothesis until checked against the code.

**Comment by ericweiner** [comment]
## Bench evidence: a saturated sweep reports a confident best with no warning (2026-09-02, LS850T, build `d7412adf`, 8-bit)

Bundle `build/SN12075-2026-09-02-132855/` (Firmware repo). The Green gain had been raised to 48 dB before a protocol scan at 13:26:44; the Green capture at 13:26:55 logged `mean=254.4 sat=97.7%`. The per-step autofocus for that step ran anyway (`autofocus.log` 13:26:50 to 13:26:55) and this is what the scorer saw, against the healthy Green sweep six seconds earlier at 13.6 dB:

| | Healthy sweep (13:26:44) | Saturated sweep (13:26:50) |
|---|---|---|
| Coarse-pass scores | 13.3M, 14.2M, 16.5M, 15.7M, 14.1M | 0, 0, 563K, -38K, 0 |
| Peak score | 16.8M | 2.1M |
| Reported best | 7049.3 um | 7032.6 um |
| Warning | none needed | none given |

The four unsaturated Green sweeps that session landed at 7047.5 to 7049.5 um; the saturated one landed 17 um away on a curve made of noise. Two guards exist and neither fired: `_mask_saturated` zeroed the 97.7% of pixels at full scale (8-bit, so the dtype threshold was correct), which is what left the noise; and the runner's degenerate-curve check `if scores.max() == 0 or scores.isna().all()` only trips when every score is zero, and two thirds of these were nonzero. Nothing in the sweep measures or logs the saturated fraction before scoring.

**What this adds to the group L3 root above:** the metric receiving a real full scale (the #319 member) is necessary but not sufficient -- with the mask working correctly the sweep still has no bound on how much of the frame may be saturated before its answer means anything. The moderate case this issue was opened for (many saturated pixels, most of the frame still sample) was not exercised that day and remains unmeasured.

No fix started, no status change.
